### PR TITLE
Support for federated authentication / enterprise SSO - take 2

### PR DESCRIPTION
**Pull request by woloski**
started from scratch because the other PR (#525) was too cluttered already after so much back and forth and merge/rebase.
- It uses [Auth10.AspNet.SimpleConfig](https://nuget.org/packages/Auth10.AspNet.SimpleConfig) nuget package that provides easier WIF integration using appSettings instead of microsoft.IdentityModel
- depends on [Microsoft.IdentityModel](https://nuget.org/packages/Microsoft.IdentityModel) nuget (which is just the WIF dll that has proper license to redist now)
- Support Windows Azure ACS and any other WS-Fed + SAML token idp
- Provides a selector like janrain that uses Windows Azure AD/ACS. This is optional (selectorEnabled = true/false in config)
- Logout will cleanup fedauth cookie
- Logic for linking identity of the federated user (being janrain or wif) moved to its own service IdentityLinker
- The feature can be turned on/off by entering empty values on the fedauth.\* appsettings

**What scenarios this enable?**
- **Enterprise chat with SSO**: Organization that wants to deploy Jabbr on the cloud or on-premises but have their employees only access with Active Directory (using ADFS or any other identity provider)
- **Chat multi-enterprise with SSO**: Multiple organizations that wants to deploy Jabbr and have a private chat system between them. They could rely on Windows Azure ACS and allow users logging in with different identity providers (even social idps like google, facebook, etc for those who don't have ADFS)
- **Chat for Office 365 orgs**: Small/mid business with a subscription with Office 365 and wants to have a private chat system for their employees and single sign on using Office 365 credentials

### CODE REVIEW

File: JabbR/default.aspx [hunk 1]
```diff
@@ -69,6 +77,24 @@
         })();
     </script>
     <% } %>
+
+    <% if (showWaadSelector)
+       { %>
```
Comment (davidfowl) [inline JabbR/default.aspx]: Fix this formatting. `<% if(showWaadSelector) { %>`

File: JabbR/Services/ApplicationSettings.cs [hunk 2]
```diff
@@ -1,4 +1,5 @@
 ﻿using System.Configuration;
+using System;
```
Comment (davidfowl) [inline JabbR/Services/ApplicationSettings.cs]: Sort usings.

File: JabbR/Hubs/Chat.cs [hunk 3]
```diff
@@ -763,6 +763,11 @@ void INotificationService.LogOut(ChatUser user, string clientId)
 
             var rooms = user.Rooms.Select(r => r.Name);
 
+            if (!string.IsNullOrEmpty(_settings.FedAuthIdentityProviderUrl))
+            {
+                FederatedIdentity.FederatedIdentityHelper.LogOff();
```
Comment (davidfowl) [inline JabbR/Hubs/Chat.cs]: What does this do?

File: JabbR/Hubs/Chat.cs [hunk 4]
```diff
@@ -763,6 +763,11 @@ void INotificationService.LogOut(ChatUser user, string clientId)
 
             var rooms = user.Rooms.Select(r => r.Name);
 
+            if (!string.IsNullOrEmpty(_settings.FedAuthIdentityProviderUrl))
```
Comment (davidfowl) [inline JabbR/Hubs/Chat.cs]: Uppercase S

File: JabbR/FederatedIdentity/Infrastructure/CustomWSFederationAuthenticationModule.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+                        // farm friendly session token handler
+                        FederatedAuthentication.ServiceConfiguration.SecurityTokenHandlers.AddOrReplace(new MachineKeySessionSecurityTokenHandler());
+                        initialized = true;
+                    }
+                }
+            }
+            
+            // do this every time the module is used
+            this.Realm = settings.Realm;
```
Comment (davidfowl) [inline JabbR/FederatedIdentity/Infrastructure/CustomWSFederationAuthenticationModule.cs]: get rid of the ``this.` prefix.

File: JabbR/FederatedIdentity/Infrastructure/CustomWSFederationAuthenticationModule.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    public partial class CustomWSFederationAuthenticationModule : WSFederationAuthenticationModule
+    {
+        private static bool initialized = false;
+        private static object locker = new object();
+
+        protected override void InitializeModule(HttpApplication context)
+        {
+            // shortcircuit registration of the module WSFederationAuthenticationModule events if fed auth is not configured
+            var settings = new FederatedIdentityConfiguration();
+            if (!string.IsNullOrEmpty(settings.IdentityProviderUrl))
```
Comment (davidfowl) [inline JabbR/FederatedIdentity/Infrastructure/CustomWSFederationAuthenticationModule.cs]: Uppercase S

File: JabbR/Hubs/Chat.cs (hunk 3 above)
Comment (woloski) [inline JabbR/Hubs/Chat.cs]: Cleanup the cookie generated by WIF. That cookie contains the user claims and will be used to populate Thread.CurrentPrincipal with a ClaimsPrincipal

File: JabbR/FederatedIdentity/Infrastructure/CustomWSFederationAuthenticationModule.Hooks.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
+﻿using System;
+using System.Linq;
+using System.Web;
+using JabbR.Services;
+using Microsoft.IdentityModel.Claims;
+using Ninject;
+using JabbR.App_Start;
+
```
Comment (davidfowl) [inline JabbR/FederatedIdentity/Infrastructure/CustomWSFederationAuthenticationModule.Hooks.cs]: Sort usings.

File: JabbR/FederatedIdentity/Infrastructure/CustomSessionAuthenticationModule.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace JabbR.FederatedIdentity.Infrastructure
+{
+    public class CustomSessionAuthenticationModule : SessionAuthenticationModule
+    {
+        protected override void InitializeModule(HttpApplication context)
+        {
+            // shortcircuit registration of the module SessionAuthenticationModule events if fed auth is not configured
+            var settings = new FederatedIdentityConfiguration();
+            if (!string.IsNullOrEmpty(settings.IdentityProviderUrl))
```
Comment (davidfowl) [inline JabbR/FederatedIdentity/Infrastructure/CustomSessionAuthenticationModule.cs]: Uppercase S

File: JabbR/FederatedIdentity/Infrastructure/CustomSessionAuthenticationModule.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+            {
+                base.InitializeModule(context);
+            }
+        }
+
+        protected override void InitializePropertiesFromConfiguration(string serviceName)
+        {
+            var settings = new FederatedIdentityConfiguration();
+
+            this.CookieHandler.RequireSsl = settings.RequiresSsl;
```
Comment (davidfowl) [inline JabbR/FederatedIdentity/Infrastructure/CustomSessionAuthenticationModule.cs]: Remove `this.`.

File: JabbR/FederatedIdentity/FederatedIdentityConfiguration.cs [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+using System.Configuration;
+using JabbR.Services;
+using JabbR.App_Start;
+using Ninject;
+
+namespace JabbR.FederatedIdentity
+{
+    public class FederatedIdentityConfiguration
+    {
+        IApplicationSettings _settings;
```
Comment (davidfowl) [inline JabbR/FederatedIdentity/FederatedIdentityConfiguration.cs]: private readonly

File: JabbR/FederatedIdentity/FederatedIdentityConfiguration.cs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+        public FederatedIdentityConfiguration()
+        {
+            this._settings = Bootstrapper.Kernel.Get<IApplicationSettings>();
```
Comment (davidfowl) [inline JabbR/FederatedIdentity/FederatedIdentityConfiguration.cs]: Why? Where did the injection go?

File: JabbR/FederatedIdentity/FederatedIdentityHelper.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+            HttpContext.Current.ApplicationInstance.CompleteRequest();
+        }
+
+
+        public static void LogOff()
+        {
+            WSFederationAuthenticationModule fam = FederatedAuthentication.WSFederationAuthenticationModule;
+            fam.SignOut(true);
+
+            HttpContext.Current.Response.Redirect("/", false);
```
Comment (davidfowl) [inline JabbR/FederatedIdentity/FederatedIdentityHelper.cs]: Why are you using HttpContext.Current?

File: JabbR/FederatedIdentity/FederatedIdentityHelper.cs [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+            var signInRequest = new SignInRequestMessage(new Uri(issuer ?? fam.Issuer), realm ?? fam.Realm)
+            {
+                AuthenticationType = fam.AuthenticationType,
+                Context = GetReturnUrl(),
+                Freshness = fam.Freshness,
+                HomeRealm = homeRealm ?? fam.HomeRealm,
+                Reply = fam.Reply
+            };
+
+            HttpContext.Current.Response.Redirect(signInRequest.WriteQueryString(), false);
```
Comment (davidfowl) [inline JabbR/FederatedIdentity/FederatedIdentityHelper.cs]: Why are you using HttpContext.Current?

File: JabbR/FederatedIdentity/Infrastructure/CustomSessionAuthenticationModule.cs [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+            var settings = new FederatedIdentityConfiguration();
+            if (!string.IsNullOrEmpty(settings.IdentityProviderUrl))
+            {
+                base.InitializeModule(context);
+            }
+        }
+
+        protected override void InitializePropertiesFromConfiguration(string serviceName)
+        {
+            var settings = new FederatedIdentityConfiguration();
```
Comment (davidfowl) [inline JabbR/FederatedIdentity/Infrastructure/CustomSessionAuthenticationModule.cs]: Why do you keep newing this up. That seems wrong. It should be injected via DI and it should be a singleton like the other settings.

**Comment by woloski** [comment]
@davidfowl is this gonna stay forever here? :)

File: JabbR/FederatedIdentity/Login.ashx.cs [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+            get
+            {
+                return false;
+            }
+        }
+
+        public void ProcessRequest(HttpContext context)
+        {
+            var settings = Bootstrapper.Kernel.Get<IApplicationSettings>();
+            if (settings.FedAuthRequiresSsl && !HttpContext.Current.Request.IsSecureConnection)
```
Comment (davidfowl) [inline JabbR/FederatedIdentity/Login.ashx.cs:39]: Use context.Request not HttpContext.Current.

File: JabbR/FederatedIdentity/Login.ashx.cs [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+            {
+                throw new InvalidOperationException("This endpoints works over TLS/SSL only");
+            }
+
+            // validate SAML token
+            var request = HttpContext.Current.Request;
```
Comment (davidfowl) [inline JabbR/FederatedIdentity/Login.ashx.cs:45]: Same here.

**Comment by davidfowl** [comment]
What's up with this?

**Comment by woloski** [comment]
death by thousand cuts :)... I need to get back to it and do a huge merge probably

**Comment by davidfowl** [comment]
Heheh, a huge rebase :wink: .

**Comment by davidfowl** [comment]
Next year? :shipit:

**Pull request closed without merging**